# MOVIES ANALYSIS WITH EDA

## BUSINESS UNDERSTANDING
### What factors should Microsoft consider when starting a Movie Studio?
Microsoft has decided to start a new Movie Studio. In order to do this they need to know the kinds of movies that are doing the best.This will help to know the type of films that microsoft should produce.EDA should be done so as to provide concrete recommendations to Microsoft.

## Data Understanding

The data for this analysis comes from;
* ### im.db
This is a SQLite database that consists of data about movies from IMDB.

* ### bom.movie_gross.csv
This is a csv file that contains data about movies from Box Office

## Importing Libraries

In [1]:
import numpy as np

#for analysis and manipulation of data
import pandas as pd

#for connecting to im.db
import sqlite3 

# for visualization
import matplotlib.pyplot as plt
import seaborn as sns

## Loading Data

In [2]:
# connecting to im.db database
conn = sqlite3.connect("im.db")

In [4]:
#load the bom data
bom_df = pd.read_csv("bom.movie_gross.csv")

## Data Understanding (Box Office)

In [5]:
# preview of first 5 rows
bom_df.head()

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010

In [6]:
# preview of the last rows
bom_df.tail()

title      studio  domestic_gross foreign_gross  \
3382                    The Quake       Magn.          6200.0           NaN   
3383  Edward II (2018 re-release)          FM          4800.0           NaN   
3384                     El Pacto        Sony          2500.0           NaN   
3385                     The Swan  Synergetic          2400.0           NaN   
3386            An Actor Prepares       Grav.          1700.0           NaN   

      year  
3382  2018  
3383  2018  
3384  2018  
3385  2018  
3386  2018

In [7]:
# Number of rows and columns
bom_df.shape

(3387, 5)

In [8]:
#Statistical summary of data
bom_df.describe()

domestic_gross         year
count    3.359000e+03  3387.000000
mean     2.874585e+07  2013.958075
std      6.698250e+07     2.478141
min      1.000000e+02  2010.000000
25%      1.200000e+05  2012.000000
50%      1.400000e+06  2014.000000
75%      2.790000e+07  2016.000000
max      9.367000e+08  2018.000000

In [9]:
# information about the columns
bom_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB


This is what we get from the above information;
* Box Office data has 5 columns namely ```title```, ```studio```, ```domestic_gross```, ```foreign_gross``` and ```year```. 
* ```title```, ```studio``` and ```foreign_gross``` columns are string ojects while ```domestic_gross``` and ```year``` are numerical objects.
* The data has a total of 3387 rows.
* The following columns have null values because they don't have all 3387 rows;
    * studio
    * domestic_gross
    * foreign_gross

## Data Cleaning(Box Office)


In [10]:
# number of rows with null values
bom_df.isnull().sum()

title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64

In [11]:
# sample rows where foreign_gross has no missing values
bom_df[bom_df["studio"].notna()].sample(5, random_state = 1)

title  studio  domestic_gross  \
571                                 Poetry    Kino        356000.0   
3343                      The Third Murder      FM         89300.0   
1850                            Camp X-Ray     IFC         13300.0   
3257  Don't Worry He Won't Get Far on Foot  Amazon       1400000.0   
1512                              Non-Stop    Uni.      92200000.0   

     foreign_gross  year  
571        1900000  2011  
3343           NaN  2018  
1850           NaN  2014  
3257       2500000  2018  
1512     130600000  2014

In [12]:
# sample rows where foreign_gross has missing values
bom_df[bom_df["studio"].isna()]

title studio  domestic_gross foreign_gross  year
210   Outside the Law (Hors-la-loi)    NaN         96900.0       3300000  2010
555         Fireflies in the Garden    NaN         70600.0       3300000  2011
933           Keith Lemon: The Film    NaN             NaN       4000000  2012
1862                 Plot for Peace    NaN          7100.0           NaN  2014
2825               Secret Superstar    NaN             NaN     122000000  2017

In [13]:
# drop rows where studio has a null value
bom_df.dropna(subset = ["studio"], inplace = True)

In [14]:
bom_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3382 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3382 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3356 non-null   float64
 3   foreign_gross   2033 non-null   object 
 4   year            3382 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 158.5+ KB


In [15]:
# percentage of missing values in domestic_gross
bom_df["domestic_gross"].isnull().sum()/len(bom_df)*100

0.768775872264932

This is a very small portion of the dataset therefore we can just drop the rows with null values.

In [16]:
bom_df.dropna(subset = ["domestic_gross"], inplace = True)
bom_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3356 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3356 non-null   object 
 1   studio          3356 non-null   object 
 2   domestic_gross  3356 non-null   float64
 3   foreign_gross   2007 non-null   object 
 4   year            3356 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 157.3+ KB


In [17]:
# percentage of missing values in foreign gross
bom_df["foreign_gross"].isnull().sum()/len(bom_df)*100

40.19666269368295

This is a significant portion of the dataset therefore we cannot just drop the rows. We can fill the missing values with either the mean or median.

In [18]:
# preview of foreign gross column
bom_df["foreign_gross"].head()

0    652000000
1    691300000
2    664300000
3    535700000
4    513900000
Name: foreign_gross, dtype: object

Looks like the datatype of foreign gross is string. So we need to change the column to a numerical datatype.

In [19]:
# cleaning the column and converting to a numerical datatype
bom_df["foreign_gross"] = [float(str(i).replace(",", "")) for i in bom_df["foreign_gross"]]
bom_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3356 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3356 non-null   object 
 1   studio          3356 non-null   object 
 2   domestic_gross  3356 non-null   float64
 3   foreign_gross   2007 non-null   float64
 4   year            3356 non-null   int64  
dtypes: float64(2), int64(1), object(2)
memory usage: 157.3+ KB


In [20]:
# mean of foreign_gross column before filling null values
bom_df["foreign_gross"].mean().round(2)

75790384.84

In [21]:
#median of foreign_gross
bom_df["foreign_gross"].median()

19400000.0

In [22]:
#checking mean after filling null values with the mean
bom_df1 = bom_df["foreign_gross"].fillna(bom_df["foreign_gross"].mean())
bom_df1.mean().round(2)

75790384.84

In [23]:
#checking mean after filling null values with median
bom_df2 = bom_df["foreign_gross"].fillna(bom_df["foreign_gross"].median())
bom_df2.mean().round(2)

53123332.05

According to the above results, filling null values with the median is not a good option because it reduces the mean. So we will fill the null values with the mean because that  does not alter the mean.

In [24]:
# replacing null values with the mean
bom_df["foreign_gross"] = bom_df["foreign_gross"].fillna(bom_df["foreign_gross"].mean())
bom_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3356 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3356 non-null   object 
 1   studio          3356 non-null   object 
 2   domestic_gross  3356 non-null   float64
 3   foreign_gross   3356 non-null   float64
 4   year            3356 non-null   int64  
dtypes: float64(2), int64(1), object(2)
memory usage: 157.3+ KB


In [25]:
#confirming there are no missing values
bom_df.isna().sum()

title             0
studio            0
domestic_gross    0
foreign_gross     0
year              0
dtype: int64

Our dataset has no missing values, therefore we can check and see if there are any duplicates.

In [26]:
#checking for duplicates
bom_df.duplicated().any()

False

Our data has no duplicates. We can then check the summary statistics after data cleaning.

In [27]:
# summary statistics of bom_df
bom_df.describe()

domestic_gross  foreign_gross         year
count    3.356000e+03   3.356000e+03  3356.000000
mean     2.877149e+07   7.579038e+07  2013.970203
std      6.700694e+07   1.068472e+08     2.479064
min      1.000000e+02   6.000000e+02  2010.000000
25%      1.200000e+05   1.220000e+07  2012.000000
50%      1.400000e+06   7.579038e+07  2014.000000
75%      2.795000e+07   7.579038e+07  2016.000000
max      9.367000e+08   9.605000e+08  2018.000000

Now that we've cleaned the Box Office dataset, we can go ahead and clean the IMDB dataset.

## Data Understanding (IMDB Data)

In [28]:
imdb_df = pd.read_sql("""
SELECT name FROM sqlite_master
WHERE type = "table";
""", conn)
imdb_df

name
0   movie_basics
1      directors
2      known_for
3     movie_akas
4  movie_ratings
5        persons
6     principals
7        writers

### PREVIEW OF THE TABLES

In [29]:
#movie_basics table
pd.read_sql("""
SELECT * FROM movie_basics;
""", conn).head()

movie_id                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

In [30]:
#directors table
pd.read_sql("""
SELECT * FROM directors;
""", conn).head()

movie_id  person_id
0  tt0285252  nm0899854
1  tt0462036  nm1940585
2  tt0835418  nm0151540
3  tt0835418  nm0151540
4  tt0878654  nm0089502

In [31]:
# known_for table
pd.read_sql("""
SELECT * FROM known_for;
""", conn).head()

person_id   movie_id
0  nm0061671  tt0837562
1  nm0061671  tt2398241
2  nm0061671  tt0844471
3  nm0061671  tt0118553
4  nm0061865  tt0896534

In [32]:
# movie_akas
pd.read_sql("""
SELECT * FROM movie_akas;
""", conn).head()

movie_id  ordering                                    title region  \
0  tt0369610        10                            Джурасик свят     BG   
1  tt0369610        11                        Jurashikku warudo     JP   
2  tt0369610        12  Jurassic World: O Mundo dos Dinossauros     BR   
3  tt0369610        13                  O Mundo dos Dinossauros     BR   
4  tt0369610        14                           Jurassic World     FR   

  language        types   attributes  is_original_title  
0       bg         None         None                0.0  
1     None  imdbDisplay         None                0.0  
2     None  imdbDisplay         None                0.0  
3     None         None  short title                0.0  
4     None  imdbDisplay         None                0.0

In [33]:
#movie_ratings
pd.read_sql("""
SELECT * FROM movie_ratings;
""", conn).head()

movie_id  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

In [34]:
# persons
pd.read_sql("""
SELECT * FROM persons;
""", conn).head()

person_id       primary_name  birth_year  death_year  \
0  nm0061671  Mary Ellen Bauder         NaN         NaN   
1  nm0061865       Joseph Bauer         NaN         NaN   
2  nm0062070         Bruce Baum         NaN         NaN   
3  nm0062195       Axel Baumann         NaN         NaN   
4  nm0062798        Pete Baxter         NaN         NaN   

                                 primary_profession  
0         miscellaneous,production_manager,producer  
1        composer,music_department,sound_department  
2                        miscellaneous,actor,writer  
3  camera_department,cinematographer,art_department  
4  production_designer,art_department,set_decorator

In [36]:
# principals table
pd.read_sql("""
SELECT * FROM principals;
""", conn).head()

movie_id  ordering  person_id  category       job        characters
0  tt0111414         1  nm0246005     actor      None       ["The Man"]
1  tt0111414         2  nm0398271  director      None              None
2  tt0111414         3  nm3739909  producer  producer              None
3  tt0323808        10  nm0059247    editor      None              None
4  tt0323808         1  nm3579312   actress      None  ["Beth Boothby"]

In [38]:
# writers table
pd.read_sql("""
SELECT * FROM writers;
""", conn).head()

movie_id  person_id
0  tt0285252  nm0899854
1  tt0438973  nm0175726
2  tt0438973  nm1802864
3  tt0462036  nm1940585
4  tt0835418  nm0310087

In this project i'll use the movie_basics table and the movie_ratings table.

## Data Cleaning (movie_basics)

In [39]:
#information on columns
moviebasics_df = pd.read_sql("""
SELECT * FROM movie_basics;
""", conn)
moviebasics_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146144 entries, 0 to 146143
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   movie_id         146144 non-null  object 
 1   primary_title    146144 non-null  object 
 2   original_title   146123 non-null  object 
 3   start_year       146144 non-null  int64  
 4   runtime_minutes  114405 non-null  float64
 5   genres           140736 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 6.7+ MB


The movie_basics table has a total of 146,144 rows.
There are columns which have missing values which are ```original_title```, ```runtime_minutes``` and ```genres```.

In [40]:
#sum of missing values
moviebasics_df.isna().sum()

movie_id               0
primary_title          0
original_title        21
start_year             0
runtime_minutes    31739
genres              5408
dtype: int64

In [42]:
# percentage of missing values in original_title column
((moviebasics_df["original_title"].isna().sum()/len(moviebasics_df)) * 100).round(2)

0.01

This is not a significant number in the dataset therefore we can just drop the rows with missing values in the original_title column.

In [45]:
#dropping rows with missing values
moviebasics_df.dropna(subset = ["original_title"], inplace = True)
moviebasics_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 146123 entries, 0 to 146143
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   movie_id         146123 non-null  object 
 1   primary_title    146123 non-null  object 
 2   original_title   146123 non-null  object 
 3   start_year       146123 non-null  int64  
 4   runtime_minutes  114401 non-null  float64
 5   genres           140734 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 7.8+ MB


In [48]:
# percentage of missing values in the runtime_minutes column
((moviebasics_df["runtime_minutes"].isna().sum()/len(moviebasics_df)) * 100).round(2)

21.71

This is a significant portion of the dataset therefore we cannot just drop the rows with missing values because it might affect our dataset. We can try filling the missing values with either the mean or the median.

In [50]:
#mean of the column before filling in the missing values
moviebasics_df["runtime_minutes"].mean().round(2)

86.19

In [52]:
# mean after filling missing values with the mean
moviebasics_df1 = moviebasics_df["runtime_minutes"].fillna(moviebasics_df["runtime_minutes"].mean())
moviebasics_df1.mean().round(2)

86.19

In [55]:
#mean after filling missing values with the median
moviebasics_df2 = moviebasics_df["runtime_minutes"].fillna(moviebasics_df["runtime_minutes"].median())
moviebasics_df2.mean().round()

86.0

According to the above output, filling the missing values with the mean is much more accurate. Filling with the median reduces the mean with ```0.19```.

In [56]:
# fill missing values with mean
moviebasics_df["runtime_minutes"] = moviebasics_df["runtime_minutes"].fillna(moviebasics_df["runtime_minutes"].mean())
moviebasics_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 146123 entries, 0 to 146143
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   movie_id         146123 non-null  object 
 1   primary_title    146123 non-null  object 
 2   original_title   146123 non-null  object 
 3   start_year       146123 non-null  int64  
 4   runtime_minutes  146123 non-null  float64
 5   genres           140734 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 7.8+ MB


In [59]:
# percentage of missing values in genres column
(moviebasics_df["genres"].isna().sum()/len(moviebasics_df) * 100).round(2)

3.69

This is not a significant portion of the dataset therefore we can just go ahead and drop the rows with null values.

In [60]:
moviebasics_df.dropna(subset = ["genres"], inplace = True)
moviebasics_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 140734 entries, 0 to 146143
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   movie_id         140734 non-null  object 
 1   primary_title    140734 non-null  object 
 2   original_title   140734 non-null  object 
 3   start_year       140734 non-null  int64  
 4   runtime_minutes  140734 non-null  float64
 5   genres           140734 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 7.5+ MB


In [62]:
#confirming there are no missing values
moviebasics_df.isna().sum()

movie_id           0
primary_title      0
original_title     0
start_year         0
runtime_minutes    0
genres             0
dtype: int64

Our dataset now has no missing values therefore we can check if there are any duplicates.

In [63]:
#checking for duplicates
moviebasics_df.duplicated().any()

False

Our data has no duplicates therefore we can go ahead and clean the next table.

## Data Cleaning (movie_ratings)

In [64]:
#information on columns
movieratings_df = pd.read_sql("""
SELECT * FROM movie_ratings;
""", conn)
movieratings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73856 entries, 0 to 73855
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   movie_id       73856 non-null  object 
 1   averagerating  73856 non-null  float64
 2   numvotes       73856 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 1.7+ MB


Our movie_ratings table has no missing values therefore we can go ahead and check if there are any duplicates.

In [66]:
movieratings_df.duplicated().any()

False